src/main/scala/mesosphere/marathon/metrics/SlidingAverageSnapshot.scala

```diff
@@ -0,0 +1,133 @@
+package mesosphere.marathon
+package metrics
+
+import java.time.Duration
+import java.util.logging.Logger
+
+import akka.Done
+import akka.actor.{ Actor, ActorRefFactory, Props }
+import kamon.Kamon
+import kamon.metric.{ Entity, EntitySnapshot }
+import com.typesafe.config.Config
+import kamon.metric.SubscriptionsDispatcher.TickMetricSnapshot
+import kamon.metric.instrument.CollectionContext
+import kamon.util.{ MapMerge, MilliTimestamp }
+import mesosphere.marathon.metrics.Metrics.subscribe
+
+/**
+  * Calculates sliding average entity snapshot
+  */
+class SlidingAverageSnapshot(val averagingWindow: Duration) {
+  private val logger: Logger = Logger.getLogger(getClass.getName)
+
+  val collectionContext: CollectionContext = Kamon.metrics.buildDefaultCollectionContext
+
+  /**
+    * The current summarised snapshot
+    */
+  private var currentSnapshot: TickMetricSnapshot = TickMetricSnapshot(MilliTimestamp.now, MilliTimestamp.now, Map.empty)
+
+  /**
+    * A ring buffer that collects the snapshots
+    */
+  private var averageRing: Array[TickMetricSnapshot] = ringFactory(Kamon.config)
+
+  /**
+    * The current index to the average ring buffer
+    */
+  private var averageIndex: Int = 0
+
+  /**
+    * Rebuilds the average ring, based on the current configuration
+    */
+  private def ringFactory(config: Config): Array[TickMetricSnapshot] = {
+    val tickInterval: Long = config.getDuration("kamon.metric.tick-interval").toMillis
+    val fullFrames: Int = (averagingWindow.toMillis.toFloat / tickInterval).ceil.toInt
+
+    // Warn for too low values
+    if (fullFrames < 3) {
+      logger.warning(s"SlidingAverageReporter applied over a very small window (${fullFrames} frames), consider" +
+        " increasing averageWindow or decreasing `kamon.metric.tick-interval`!")
+    }
+
+    // Create an array with that many free slots, as the full ticks that fit within the duration requested
+    logger.info(s"Initialized sliding average reporter with ${fullFrames} frames")
+    Array.fill(fullFrames){null}
+  }
+
+  /**
+    * Combine all the metrics in the ring buffer and return an
+    * @return
+    */
+  private def combineRingMetrics: TickMetricSnapshot = {
+    var rangeFrom: MilliTimestamp = MilliTimestamp.now
+    var rangeTo: MilliTimestamp = MilliTimestamp.now
+    var combined: Option[Map[Entity, EntitySnapshot]] = None
+
+    // Start processing all the values of the ring buffer in incrementing index
+    // starting fro the oldest (the next value from the current index) to the
+    // newest (the current index).
+    val maxItems: Int = averageRing.length
```

Comment:
I guess with the Apache ring buffer this is something like

```scala
buffer.toArray.sortBy(...).fold() { (accumulated, next) =>
   MapMerge.Syntax(accumulated).merge(next ....)
}
```

what do you think?

Reply:
So on every tick we sort the list? That doesn't sound attractive 😄

Reply:
You don't have to. I was just not sure why you need max. I thought you might want to accumulate in order. If the number of ticks is low sorting will probably not be a big impact.

Reply:
The list is always sorted by date. That's the part that I am interested. I guess `CircularFifoBuffer` does not sort anything.